javatests/arcs/core/entity/TypeSlicingTest.kt

```diff
@@ -0,0 +1,189 @@
+    handles.l.store(bottom)
+    handles.r.store(right)
+    handles.r.store(bottom)
+  }
+}
+
+// Reader collectes the sliced writes from Slicer for verification in the unit test.
+class Reader : AbstractReader() {
+  override fun onStart() {
+    handles.t.onUpdate { topReceived.add(requireNotNull(it.new)) }
+    handles.l.onUpdate { leftReceived.add(requireNotNull(it.new)) }
+    handles.r.onUpdate { rightReceived.add(requireNotNull(it.new)) }
+  }
+
+  override fun onUpdate() {
+    if (--expected == 0) latch.complete()
+  }
+
+  companion object {
+    var expected = 0
+    val latch = Job() // Single job works since only one test requires it.
+    val topReceived = mutableListOf<Reader_T>()
+    val leftReceived = mutableListOf<Reader_L>()
+    val rightReceived = mutableListOf<Reader_R>()
+  }
+}
+
+@RunWith(JUnit4::class)
+@OptIn(ExperimentalCoroutinesApi::class)
+class TypeSlicingTest {
+
+  @Test
+  fun particlesCanWriteToSliceInterfaces() = runTest {
+    RamDisk.clear()
+    DriverAndKeyConfigurator.configure(null)
+    val schedulerProvider = SimpleSchedulerProvider(coroutineContext)
+    val host = TestHost(
+      schedulerProvider,
+      ::Reader.toRegistration(),
+      ::Slicer.toRegistration(),
+      ::Writer.toRegistration()
+    )
+    val hostRegistry = ExplicitHostRegistry().apply { registerHost(host) }
+    val allocator = Allocator.createNonSerializing(hostRegistry, TestCoroutineScope())
+    val arc = allocator.startArcForPlan(TypeSlicingTestPlan).waitForStart()
+
+    Reader.latch.join()
+
+    assertThat(Reader.topReceived.map { it.a }).isEqualTo(listOf("left", "right", "bottom"))
+
+    assertThat(Reader.leftReceived.map { "${it.a}:${it.b}" })
+      .isEqualTo(listOf("left:2.3", "bottom:8.0"))
+
+    assertThat(Reader.rightReceived.map { "${it.a}:${it.c}" })
+      .isEqualTo(listOf("right:[]", "bottom:[5, 9]"))
+
+    arc.stop()
+    arc.waitForStop()
+    schedulerProvider.cancelAll()
+  }
+
+  @Test
+  fun castingBetweenRelatedSchemas() {
+    // Re-use the type lattice from Reader to check direct casts across the entire lattice.
+    checkCasts(Reader_T("abc"), true, false, false, false)
+    checkCasts(Reader_L("abc", 5.7), true, true, false, false)
+    checkCasts(Reader_R("abc", setOf(4, 6)), true, false, true, false)
+    checkCasts(Reader_B("abc", 5.7, setOf(4, 6), true), true, true, true, true)
+
+    // An unrelated type should fail all casts.
+    checkCasts(Reader_X(5.7), false, false, false, false)
+  }
+
+  fun checkCasts(x: Any, isTop: Boolean, isLeft: Boolean, isRight: Boolean, isBottom: Boolean) {
+    try {
+      (x as Reader_T_Slice).let {
+        assertThat(it.a).isEqualTo("abc")
+      }
+      assertWithMessage("Cast of $x to Reader_T_Slice succeeded but should have failed")
+        .that(isTop).isTrue()
+    } catch (e: ClassCastException) {
+      assertWithMessage("Cast of $x to Reader_T_Slice failed but should have succeeded")
+        .that(isTop).isFalse()
+    }
```

Comment:
Optional: This might be a bit clearer as:

```
if (isTop) {
  assertFailsWith<ClassCastException> { x as Reader_T_Slice }
} else {
  x as Reader_T_Slice
}
```

Same below.

Reply:
Nice! That's much clearer.